pkg/packagekit/updates.jsx

```diff
@@ -352,6 +363,100 @@ const UpdatesList = ({ updates }) => {
     );
 };
 
+class RestartServices extends React.Component {
+    constructor(props) {
+        super(props);
+        this.state = {
+            dialogError: undefined,
+            restartInProgress: false,
+        };
+
+        this.dialogErrorSet = this.dialogErrorSet.bind(this);
+        this.dialogErrorDismiss = this.dialogErrorDismiss.bind(this);
+        this.restart = this.restart.bind(this);
+    }
+
+    dialogErrorSet(text, detail) {
+        this.setState({ dialogError: text, dialogErrorDetail: detail });
+    }
+
+    dialogErrorDismiss() {
+        this.setState({ dialogError: undefined });
+    }
+
+    restart() {
+        // make sure cockpit package is the last to restart
+        const daemons = this.props.daemons.sort((a, b) => {
+            if (a.includes("cockpit") && b.includes("cockpit"))
+                return 0;
+            if (a.includes("cockpit"))
+                return 1;
+            return a.localeCompare(b);
+        });
+        const restarts = daemons.map(service => cockpit.spawn(["systemctl", "restart", service + ".service"], { superuser: "required", err: "message" }));
+        this.setState({ restartInProgress: true });
+        Promise.all(restarts)
+                .then(() => {
+                    this.props.onValueChanged({ tracerPackages: { reboot: [], daemons: [], manual: [] } });
+                    if (this.props.state === "updateSuccess")
+                        this.props.loadUpdates();
+                    this.setState({ restartInProgress: false });
+                    this.props.close();
+                })
+                .catch(ex => {
+                    this.dialogErrorSet(_("Failed to restart service"), ex.message);
+                    // call Tracer again to see what services remain
+                    this.props.callTracer(null);
+                });
+    }
+
+    render() {
+        let body;
+        if (this.props.tracerRunning) {
+            body = (<>
+                <div className="spinner spinner-xs spinner-inline" />
+                <p>{_("Reloading the state of remaining services")}</p>
+            </>);
+        } else if (this.props.daemons.length > 0) {
+            body = (<>
+                {cockpit.ngettext("The following service will be restarted:", "The following services will be restarted:", this.props.daemons.length)}
+                <TwoColumnContent list={this.props.daemons} flexClassName="restart-services-modal-body" />
+            </>);
+        }
```

Comment:
It seems it could happen that tracer is finished, but there are no daemons to restart?

Reply:
Yes, I added that tweak after our discussions about calling tracer again when service restart fails (here https://github.com/cockpit-project/cockpit/pull/14581#discussion_r568359285).
If tracer wants to restart some services, and almost all of them restart correctly, but one service fails (and that one service after failing doesn't run any process anymore, therefore it doesn't use any outdated files anymore, therefore it's technically no longer a service which needs a restart, therefore tracer doesn't pick it up), it will result in a dialog being empty.
Yet the dialog still needs to be left open to show error message so user can read it.

Reply:
Makes sense, thanks!